### PR TITLE
Add support for PubSub subscriptions with binary channel names and handlers

### PR DESCRIPTION
**Pull request by petyaslavova**
## Summary

This PR adds a public `Subscription(name, handler=None)` value type for PubSub subscriptions and accepts it positionally in `subscribe`, `psubscribe`, and `ssubscribe` across sync, async, and cluster PubSub paths.

The new input form removes the kwargs-only limitation for handler-backed subscriptions. 
Since Python keyword argument keys must be text strings, users previously could not attach handlers to arbitrary binary channel, pattern, or shard channel names. 
`Subscription` keeps the existing APIs intact while adding a binary-safe way to bind handlers.

Reconnect replay now preserves handler-backed binary subscriptions by replaying them as positional `Subscription` objects instead of decoding names into kwargs. 
Cluster sharded PubSub forwarding and shard migration paths were updated similarly, so handler associations survive reconnects and slot movement without requiring text-decodable channel names.

Docs now show the new `Subscription` form, mixed positional usage, and compatibility with the existing kwargs handler style. 
Type hints were updated for the affected subscribe methods, and `Subscription` is re-exported from both `redis` and `redis.asyncio`.

Tests cover sync and async binary channel, pattern, and shard channel handlers, reconnect replay with handlers, and cluster shard routing/migration behavior. 

---

> [!NOTE]
> **Medium Risk**
> Touches PubSub subscribe/resubscribe and cluster shard routing paths across sync+async clients; behavior changes around handler attachment and reconnect replay could affect subscription state in edge cases, though changes are well-covered by new tests.
> 
> **Overview**
> Adds a public `Subscription(name, handler=None)` value type and extends `subscribe`/`psubscribe`/`ssubscribe` (sync, asyncio, cluster, and multidb) to accept handler-backed subscriptions positionally, enabling callbacks for *binary* channel/pattern/shard names that can’t be expressed as kwargs.
> 
> Refactors PubSub argument parsing and reconnect replay to use new helper utilities (`parse_pubsub_subscriptions`, `pubsub_subscription_args`) so handler-backed subscriptions are re-applied without decoding names into keyword keys; updates cluster shard subscription forwarding/migration to pass `Subscription` objects instead of decoded kwargs.
> 
> Re-exports `Subscription` from `redis` and `redis.asyncio`, updates docs with mixed-form examples, and adds sync/async tests covering binary handler subscriptions, reconnect resubscribe, and cluster shard migration behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8e129d23c274742d5bb70ad5b3f9ad621d972941. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: redis/commands/helpers.py [hunk 1]
```diff
@@ -37,6 +37,29 @@ def list_or_args(keys: KeysT, args: Tuple[KeyT, ...]) -> List[KeyT]:
...
+
+def pubsub_subscription_args(
+    subscriptions: Mapping[ChannelT, PubSubHandler | None],
+) -> list[ChannelT | Subscription]:
+    return [
+        channel if handler is None else Subscription(channel, handler)
+        for channel, handler in subscriptions.items()
+    ]
+
+
```
[the remark this answers is not in the document]
Comment (petyaslavova) [reply redis/commands/helpers.py:57]: fixed

**Review by vladvildanov — APPROVED** [review]
